Firmware Request for MSI Prestige 15 A10SC

### Laptop model

MSI Prestige 15 A10SC

### EC firmware version

16S3EMS1.103

### EC memory dump

```text
Sound Off and Mic Off

00000000  00 80 00 00 00 00 00 00  00 00 00 00 00 00 00 00
00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
00000020  00 00 00 00 00 00 00 00  00 00 20 80 fb 06 bf 0b
00000030  03 01 00 05 51 0a 05 00  a0 14 60 3b 71 02 c0 00
00000040  18 15 4f 00 f8 0d 00 00  f5 0a df 3c d6 0b f8 43
00000050  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
00000060  00 00 00 00 00 00 00 00  31 00 32 3a 41 46 5a 5f
00000070  64 00 00 3a 41 48 50 55  64 00 03 05 05 05 05 03
00000080  00 00 32 3c 46 52 5a 5d  64 00 00 32 41 48 50 55
00000090  64 00 03 05 05 06 03 03  02 00 6e 02 00 66 00 00
000000a0  31 36 53 33 45 4d 53 31  2e 31 30 33 31 31 30 37
000000b0  32 30 31 39 31 37 3a 30  34 3a 33 30 f8 0d 40 40
000000c0  00 07 23 00 00 00 00 00  00 5c 00 00 00 00 14 00
000000d0  00 00 40 00 70 00 00 00  00 00 00 80 00 00 00 00
000000e0  e2 02 00 03 10 00 00 00  00 00 00 00 00 9a 00 bc
000000f0  00 00 80 81 8d 0f 05 00  00 00 00 00 00 00 00 00

Sound On and Mic Off

00000000  00 80 00 00 00 00 00 00  00 00 00 00 00 00 00 00
00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
00000020  00 00 00 00 00 00 00 00  00 00 20 80 fb 06 bf 0b
00000030  03 01 00 05 51 0a 05 00  a0 14 60 3b 71 02 c0 00
00000040  18 15 4f 00 f8 0d 00 00  f5 0a df 3c d7 0b f8 43
00000050  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
00000060  00 00 00 00 00 00 00 00  32 00 32 3a 41 46 5a 5f
00000070  64 00 00 3a 41 48 50 55  64 00 03 05 05 05 05 03
00000080  00 00 32 3c 46 52 5a 5d  64 00 00 32 41 48 50 55
00000090  64 00 03 05 05 06 03 03  02 00 6e 02 00 66 00 00
000000a0  31 36 53 33 45 4d 53 31  2e 31 30 33 31 31 30 37
000000b0  32 30 31 39 31 37 3a 30  34 3a 33 30 f8 0d 40 40
000000c0  00 07 23 00 00 00 00 00  00 84 00 00 00 00 14 00
000000d0  00 00 40 00 70 00 00 00  00 00 00 80 00 00 00 00
000000e0  e2 02 00 03 10 00 00 00  00 00 00 00 00 9a 00 bc
000000f0  00 00 80 81 8d 0f 05 00  00 00 00 00 00 00 00 00

Sound Off and Mic On

00000000  00 80 00 00 00 00 00 00  00 00 00 00 00 00 00 00
00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
00000020  00 00 00 00 00 00 00 00  00 00 20 80 fb 06 bf 0b
00000030  03 01 00 0d 51 0a 05 00  a0 14 60 3b 71 02 c0 00
00000040  18 15 4f 00 f8 0d 00 00  f5 0a df 3c d6 0b f8 43
00000050  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
00000060  00 00 00 00 00 00 00 00  32 00 32 3a 41 46 5a 5f
00000070  64 3a 00 3a 41 48 50 55  64 00 03 05 05 05 05 03
00000080  00 00 32 3c 46 52 5a 5d  64 00 00 32 41 48 50 55
00000090  64 00 03 05 05 06 03 03  02 00 6e 02 00 66 00 00
000000a0  31 36 53 33 45 4d 53 31  2e 31 30 33 31 31 30 37
000000b0  32 30 31 39 31 37 3a 30  34 3a 33 30 f8 0d 40 40
000000c0  00 07 23 00 00 00 00 00  00 7f 00 00 00 8e 14 00
000000d0  00 00 40 00 70 00 00 00  00 00 00 80 00 00 00 00
000000e0  e2 02 00 03 10 00 00 00  00 00 00 00 00 9a 00 bc
000000f0  00 00 80 81 8d 0f 05 00  00 00 00 00 00 00 00 00
```


### GPU

Nvidia

### Is your keyboard RGB?

No (single color)

### Additional context

Setup: NixOS 24.05, kernel 6.10.6, Nvidia driver 560.35.03

By default, FN keys work, including ESC and POWER LEDs.

By using [MControlCenter](https://github.com/dmitry-s93/MControlCenter) I have the following extra functionalities:
- Fan control and cooler boost
- Battery threshold (50-60, 70-80, 90-100, and custom limit)

The following (relevant) thing does not work: F1/F5 LEDs (sound and mic).

With NixOS I cannot install normally the driver due to immutability, but I was able to load it in the configurations.
However, when trying to load the driver as debug mode, the following errors emerge:
- insmod /run/booted-system/kernel-modules/lib/modules/6.10.6/kernel/drivers/platform/x86/msi-ec.ko.xz debug=1                                                                       
- modprobe: ERROR: could not insert 'msi_ec': Operation not supported

Dmesg shows:
- msi_ec: unknown parameter 'debug' ignored
- msi_ec: Firmware version is not supported: '16S3EMS1.103'

Probably some older version is loaded on the kernel that does not have debug mode implemented.

Hi, thank you for the fast reply. 

However, I cannot install the driver due to the following errors:
- make[1]: Entering directory '/home/pleiades/msi-ec'
- make[1]: *** /lib/modules/6.10.6/build: No such file or directory.  Stop.
- make[1]: Leaving directory '/home/pleiades/msi-ec'
- make: *** [Makefile:16: modules] Error 2

I could not find any kind of NixOS configuration that allowed me to install the latest version.

--- a/msi-ec.c
+++ b/msi-ec.c
@@ -473,6 +473,7 @@ static struct msi_ec_conf CONF5 __initdata = {
 };
 
 static const char *ALLOWED_FW_6[] __initconst = {
+	"1541EMS1.113", // GE66 Raider 10SF
 	"1542EMS1.102", // GP66 Leopard 10UG / 10UE / 10UH
 	"1542EMS1.104",
 	NULL
@@ -483,7 +484,7 @@ static struct msi_ec_conf CONF6 __initdata = {
 	.charge_control_address = 0xef,
 	.webcam = {
 		.address       = 0x2e,
-		.block_address = MSI_EC_ADDR_UNSUPP,
+		.block_address = 0x2f,
 		.bit           = 1,
 	},
 	.fn_win_swap = {
@@ -506,7 +507,7 @@ static struct msi_ec_conf CONF6 __initdata = {
 		},
 	},
 	.super_battery = {
-		.address = 0xd5,
+		.address = MSI_EC_ADDR_UNKNOWN,
 		.mask    = 0x0f,
 	},
 	.fan_mode = {
@@ -1032,6 +1033,7 @@ static const char *ALLOWED_FW_14[] __initconst = {
 	"17L2EMS1.108", // Katana 17 B11UCX, Katana GF76 11UC
 	"17L3EMS1.106", // Crosshair 17 B12UGZ
 	"17L3EMS1.109", // Katana GF76 12UG
+	"17L4EMS1.112", // Katana GF76 12UC
 	NULL
 };
 
@@ -1521,6 +1523,7 @@ static struct msi_ec_conf CONF20 __initdata = {
 };
 
 static const char *ALLOWED_FW_21[] __initconst = {
+	"16R1EMS1.105", // GF63 8RC-249
 	"16R3EMS1.100", // GF63 Thin 9SC
 	"16R3EMS1.102",
 	"16R3EMS1.104",
@@ -2221,6 +2224,8 @@ static struct msi_ec_conf CONF30 __initdata = {
 };
 
 static const char *ALLOWED_FW_31[] __initconst = {
+	"16Q2EMS1.107", // GS65 Stealth Thin 8RF
+	"16Q2EMS1.T40",
 	"16Q4EMS1.110", // GS65 Stealth
 	NULL
 };
@@ -2230,7 +2235,7 @@ static struct msi_ec_conf CONF31 __initdata = {
 	.charge_control_address = 0xef,
 	.webcam = {
 		.address       = 0x2e,
-		.block_address = MSI_EC_ADDR_UNSUPP,
+		.block_address = 0x2f,
 		.bit           = 1,
 	},
 	.fn_win_swap = {
@@ -2258,8 +2263,8 @@ static struct msi_ec_conf CONF31 __initdata = {
 	.fan_mode = {
 		.address = 0xf4,
 		.modes = {
-			{ FM_BASIC_NAME,    0x4c },
 			{ FM_AUTO_NAME,     0x0c },
+			{ FM_BASIC_NAME,    0x4c },
 			{ FM_ADVANCED_NAME, 0x8c },
 			MSI_EC_MODE_NULL
 		},
@@ -2425,6 +2430,7 @@ static struct msi_ec_conf CONF33 __initdata = {
 };
 
 static const char *ALLOWED_FW_34[] __initconst = {
+	"14C4EMS1.120", // Prestige 14 A11SCX
 	"14C6EMS1.109", // Prestige 14 Evo A12M
 	NULL
 };
@@ -2440,7 +2446,7 @@ static struct msi_ec_conf CONF34 __initdata = {
 	.fn_win_swap = {
 		.address = 0xe8,
 		.bit     = 4,
-		.invert  = false,
+		.invert  = true,
 	},
 	.cooler_boost = {
 		.address = 0x98,
@@ -2638,6 +2644,8 @@ static const char *ALLOWED_FW_37[] __initconst = {
 	"15M1IMS1.110", // Vector GP68 HX 13V
 	"15M1IMS1.113", // Vector GP68 HX 12V
 	"15M3EMS1.105", // Vector 16 HX AI A2XWHG
+	"15M3EMS1.106"
+	"15M3EMS1.107",
 	NULL
 };
 
@@ -3484,7 +3492,8 @@ static struct msi_ec_conf CONF48 __initdata = {
 static const char *ALLOWED_FW_49[] __initconst = {
 	"16R4EMS1.101", // GF63 Thin 10SCX(R) / 10SCS(R)
 	"16R4EMS1.102",
-	"16R5EMS1.102", // GF63 Thin 10U(C/D) / 10SC
+	"16R5EMS1.101", // GF63 Thin 10U(C/D) / 10SC
+	"16R5EMS1.102",
 	NULL
 };
 

--- a/msi-ec.c
+++ b/msi-ec.c
@@ -124,12 +124,12 @@ static struct msi_ec_conf CONF0 __initdata = {
 };
 
 static const char *ALLOWED_FW_1[] __initconst = {
-	"17F4EMS2.100", // GF75 Thin 9SCSR
-	"17F3EMS2.103", // GF75 Thin 10SER
 	"17F2EMS1.103", // GF75 Thin 9SC
 	"17F2EMS1.104",
 	"17F2EMS1.106",
 	"17F2EMS1.107",
+	"17F3EMS2.103", // GF75 Thin 10SER
+	"17F4EMS2.100", // GF75 Thin 9SCSR
 	NULL
 };
 
@@ -1030,6 +1030,8 @@ static const char *ALLOWED_FW_14[] __initconst = {
 	"17L1EMS1.106", // Katana GF76 11UG
 	"17L1EMS1.107", // Katana GF76 11UE
 	"17L2EMS1.108", // Katana 17 B11UCX, Katana GF76 11UC
+	"17L3EMS1.106", // Crosshair 17 B12UGZ
+	"17L3EMS1.109", // Katana GF76 12UG
 	NULL
 };
 
@@ -1064,9 +1066,8 @@ static struct msi_ec_conf CONF14 __initdata = {
 		},
 	},
 	.super_battery = {
-		.address = MSI_EC_ADDR_UNSUPP, // enabled by Low Performance Level
-		// .address = 0xeb, // states: 0x00 || 0x0f
-		.mask    = 0x0f,
+		.address = 0xeb,
+		.mask    = 0x0f
 	},
 	.fan_mode = {
 		.address = 0xd4,
@@ -2634,7 +2635,9 @@ static struct msi_ec_conf CONF36 __initdata = {
 };
 
 static const char *ALLOWED_FW_37[] __initconst = {
+	"15M1IMS1.110", // Vector GP68 HX 13V
 	"15M1IMS1.113", // Vector GP68 HX 12V
+	"15M3EMS1.105", // Vector 16 HX AI A2XWHG
 	NULL
 };
 
@@ -3549,6 +3552,7 @@ static struct msi_ec_conf CONF49 __initdata = {
 };
 
 static const char *ALLOWED_FW_50[] __initconst = {
+	"1584EMS1.104",	// Katana GF66 12UD
 	"1584EMS1.112", // Katana GF66 12UC
 	"1583EMS1.110", // Pulse  GL66 12UGK / Crosshair 15 B12UEZ / Katana GF66-12UG
 	NULL
@@ -3617,7 +3621,8 @@ static struct msi_ec_conf CONF50 __initdata = {
 };
 
 static const char *ALLOWED_FW_51[] __initconst = {
-	"158MEMS1.101", // Bravo 15 B5ED
+	"158MEMS1.100", // Bravo 15 B5ED
+	"158MEMS1.101",
 	"158KEMS1.104", // Bravo 15 B5DD
 	"158KEMS1.106",
 	"158KEMS1.109",
@@ -3690,6 +3695,7 @@ static const char *ALLOWED_FW_52[] __initconst = {
 	"16V1EMS1.109", // GS66 Stealth 10SFS
 	"16V1EMS1.116",
 	"16V1EMS1.118", // GS66 Stealth 10SE
+	"16V2EMS1.104", // Creator 15 A10SD
 	"16V3EMS1.106", // GS66 Stealth 10UE
 	NULL
 };
@@ -3741,9 +3747,9 @@ static struct msi_ec_conf CONF52 __initdata = {
 		.rt_temp_address      = 0x80,
 		.rt_fan_speed_address = 0x89,
 	},
-	.leds = {
-		.micmute_led_address = MSI_EC_ADDR_UNSUPP,
-		.mute_led_address    = MSI_EC_ADDR_UNSUPP,
+	.leds = { // enabled for Creator 15 A10SD
+		.micmute_led_address = 0x2b,
+		.mute_led_address    = 0x2c,
 		.bit                 = 2,
 	},
 	.kbd_bl = {
@@ -3823,6 +3829,7 @@ static struct msi_ec_conf CONF53 __initdata = {
 static const char *ALLOWED_FW_54[] __initconst = {
 	"16R8IMS2.112", // Thin 15 B12UCX / B12VE
 	"16R8IMS2.117",
+	"16RKIMS1.110", // Thin A15 B7VF
 	NULL
 };
 
@@ -3831,7 +3838,7 @@ static struct msi_ec_conf CONF54 __initdata = {
 	.charge_control_address = 0xd7,
 	.webcam = { // not present in app, but ec supports
 		.address       = 0x2e,
-		.block_address = MSI_EC_ADDR_UNSUPP,
+		.block_address = 0x2f,
 		.bit           = 1,
 	},
 	.fn_win_swap = {
@@ -3889,6 +3896,7 @@ static struct msi_ec_conf CONF54 __initdata = {
 static const char *ALLOWED_FW_55[] __initconst = {
 	"17G1EMS2.106", // P75  CREATOR 9SG
 	"17G1EMS1.107", // GS75 Stealth 9SF
+	"17G3EMS1.115", // GS75 Stealth 10SF
 	NULL
 };
 

--- a/msi-ec.c
+++ b/msi-ec.c
@@ -1174,6 +1174,7 @@ static struct msi_ec_conf CONF15 __initdata = {
 static const char *ALLOWED_FW_16[] __initconst = {
 	"155LEMS1.105", // Modern 15 A5M
 	"155LEMS1.106",
+	"1551EMS1.106", // Modern 15 A10M
 	NULL
 };
 
@@ -1188,7 +1189,7 @@ static struct msi_ec_conf CONF16 __initdata = {
 	.fn_win_swap = {
 		.address = 0xbf,
 		.bit     = 4,
-		.invert  = false,
+		.invert  = true,
 	},
 	.cooler_boost = {
 		.address = 0x98,
@@ -1199,7 +1200,7 @@ static struct msi_ec_conf CONF16 __initdata = {
 		.modes = {
 			{ SM_ECO_NAME,     0xc2 },
 			{ SM_COMFORT_NAME, 0xc1 },
-			{ SM_SPORT_NAME,   0xc0 },
+			{ SM_TURBO_NAME,   0xc0 },
 			MSI_EC_MODE_NULL
 		},
 	},
@@ -3128,6 +3129,8 @@ static struct msi_ec_conf CONF43 __initdata = {
 };
 
 static const char *ALLOWED_FW_44[] __initconst = {
+	"17L5EMS1.111", // Pulse/Katana 17 B13V/GK
+	"17L5EMS2.115", // Katana 17 B12VEK
 	"17LNIMS1.10E", // Bravo 17 C7VE
 	"17LNIMS1.505", // Katana A17 AI B8VF
 	NULL
@@ -3142,7 +3145,7 @@ static struct msi_ec_conf CONF44 __initdata = {
 	// },
 	.webcam = {
 		.address       = 0x2e,
-		.block_address = MSI_EC_ADDR_UNSUPP, // not in MSI app
+		.block_address = 0x2f,
 		.bit           = 1,
 	},
 	.fn_win_swap = {
@@ -3157,8 +3160,8 @@ static struct msi_ec_conf CONF44 __initdata = {
 	.shift_mode = {
 		.address = 0xd2,
 		.modes = {
-			{ SM_COMFORT_NAME, 0xc1 }, // Silent / Balanced / AI
 			{ SM_ECO_NAME,     0xc2 }, // Super Battery
+			{ SM_COMFORT_NAME, 0xc1 }, // Silent / Balanced / AI
 			{ SM_TURBO_NAME,   0xc4 }, // Performance
 			MSI_EC_MODE_NULL
 		},

--- a/msi-ec.c
+++ b/msi-ec.c
@@ -58,6 +58,7 @@ static const char *ALLOWED_FW_0[] __initconst = {
 	"14C1EMS1.012", // Prestige 14 A10SC
 	"14C1EMS1.101",
 	"14C1EMS1.102",
+	"16S3EMS1.103", // Prestige 15 A10SC
 	NULL
 };
 
@@ -2782,11 +2783,15 @@ static struct msi_ec_conf CONF38 __initdata = {
 		.max_state        = 3,
 	},
 };
+
 static const char *ALLOWED_FW_39[] __initconst = {
+	"16R7IMS1.104", // Thin GF63 12HW
 	"16R8IMS1.107", // Thin GF63 12VE
 	"16R8IMS1.108", // Thin GF63 12UCX
 	"16R8IMS1.111", // Thin GF63 12V(E/F)
 	"16R8IMS1.117", // Thin GF63 12UC
+	"16R8IMS2.112", // Thin 15 B12UCX / B12VE
+	"16R8IMS2.117",
 	NULL
 };
 
@@ -2810,9 +2815,9 @@ static struct msi_ec_conf CONF39 __initdata = {
 	.shift_mode = {
 		.address = 0xd2,
 		.modes = {
-			{ SM_ECO_NAME,     0xc2 },
-			{ SM_COMFORT_NAME, 0xc1 },
-			{ SM_TURBO_NAME,   0xc4 },
+			{ SM_ECO_NAME,     0xc2 }, // eco-silent + super battery
+			{ SM_COMFORT_NAME, 0xc1 }, // balanced
+			{ SM_TURBO_NAME,   0xc4 }, // perf
 			MSI_EC_MODE_NULL
 		},
 	},
@@ -2823,13 +2828,13 @@ static struct msi_ec_conf CONF39 __initdata = {
 	.fan_mode = {
 		.address = 0xd4,
 		.modes = {
-			{ FM_AUTO_NAME,      0x0d},
-			{ FM_SILENT_NAME,    0x1d},
-			{ FM_ADVANCED_NAME,  0x8d},
+			{ FM_AUTO_NAME,     0x0d },
+			{ FM_SILENT_NAME,   0x1d }, // not used in Eco-silent
+			{ FM_ADVANCED_NAME, 0x8d },
 			MSI_EC_MODE_NULL
 		},
 	},
-	.cpu = {
+	.cpu = { // single fan
 		.rt_temp_address      = 0x68,
 		.rt_fan_speed_address = 0x71,
 	},
@@ -3565,7 +3570,7 @@ static struct msi_ec_conf CONF49 __initdata = {
 };
 
 static const char *ALLOWED_FW_50[] __initconst = {
-	"1584EMS1.104",	// Katana GF66 12UD
+	"1584EMS1.104",// Katana GF66 12UD
 	"1584EMS1.112", // Katana GF66 12UC
 	"1583EMS1.110", // Pulse  GL66 12UGK / Crosshair 15 B12UEZ / Katana GF66-12UG
 	NULL
@@ -3708,7 +3713,6 @@ static const char *ALLOWED_FW_52[] __initconst = {
 	"16V1EMS1.109", // GS66 Stealth 10SFS
 	"16V1EMS1.116",
 	"16V1EMS1.118", // GS66 Stealth 10SE
-	"16V2EMS1.104", // Creator 15 A10SD
 	"16V3EMS1.106", // GS66 Stealth 10UE
 	NULL
 };

--- a/msi-ec.c
+++ b/msi-ec.c
@@ -1175,6 +1175,7 @@ static const char *ALLOWED_FW_16[] __initconst = {
 	"155LEMS1.105", // Modern 15 A5M
 	"155LEMS1.106",
 	"1551EMS1.106", // Modern 15 A10M
+	"1551EMS1.107",
 	NULL
 };
 
@@ -2645,7 +2646,7 @@ static const char *ALLOWED_FW_37[] __initconst = {
 	"15M1IMS1.110", // Vector GP68 HX 13V
 	"15M1IMS1.113", // Vector GP68 HX 12V
 	"15M3EMS1.105", // Vector 16 HX AI A2XWHG
-	"15M3EMS1.106"
+	"15M3EMS1.106",
 	"15M3EMS1.107",
 	NULL
 };

